File: quant-black-box/backend/app/routers/bs.py
from fastapi import APIRouter, HTTPException

from .. import engine
from ..schemas import BsParams, BsRequest, BsResponse, GridPoint, GridResponse

router = APIRouter()


def build_grid(req: BsRequest) -> GridResponse:
    n = req.grid.n
    x0, x1 = req.K * 0.4, req.K * 1.6
    y0, y1 = 0.05, 1.6
    points: list[GridPoint] = []
    for i in range(n):
        x = x0 + (x1 - x0) * i / (n - 1)
        for j in range(n):
            t = y0 + (y1 - y0) * j / (n - 1)
            m = engine.bs(x, req.K, t, req.r, req.sig)
            if m is None:
                continue
            if req.grid.metric == "price":
                z = m.C if req.opt == "call" else m.P
            else:
                z = m.delta if req.opt == "call" else m.deltaP
            points.append(GridPoint(x=x, y=t, z=z))
    return GridResponse(
        n=n,
        metric=req.grid.metric,
        points=points,
        shape=[n, n],
        x_range=[x0, x1],
        y_range=[y0, y1],
    )


@router.post("", response_model=BsResponse)
def price(req: BsRequest) -> BsResponse:
    m = engine.bs(req.S0, req.K, req.T, req.r, req.sig)
    if m is None:
        raise HTTPException(
            status_code=422, detail="Degenerate inputs: S0, K, T and sig must be positive"
        )
    grid = build_grid(req) if req.grid is not None else None
    return BsResponse(
        opt=req.opt,
        params=BsParams(S0=req.S0, K=req.K, T=req.T, r=req.r, sig=req.sig),
        C=m.C,
        P=m.P,
        delta=m.delta,
        deltaP=m.deltaP,
        gamma=m.gamma,
        vega=m.vega,
        theta=m.theta,
        rho=m.rho,
        d1=m.d1,
        d2=m.d2,
        grid=grid,
    )

File: quant-black-box/server/src/routes/bs.ts
import { Router } from 'express';
import { z } from 'zod';
import { bs, normCdf, normPdf } from '../engine.js';

const router = Router();

const gridSchema = z.object({
  n: z.coerce.number().min(2).max(120).default(26),
  metric: z.enum(['price', 'delta']).default('price'),
});

const bsSchema = z.object({
  S0: z.coerce.number().gt(0),
  K: z.coerce.number().gt(0),
  T: z.coerce.number().gt(0),
  r: z.coerce.number(),
  sig: z.coerce.number().gt(0),
  opt: z.enum(['call', 'put']).default('call'),
  grid: gridSchema.nullish(),
});

router.post('/', (req, res) => {
  const parsed = bsSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(422).json({ ok: false, error: 'Invalid parameters', details: parsed.error.flatten() });
    return;
  }
  const { S0, K, T, r, sig, opt, grid } = parsed.data;
  const m = bs(S0, K, T, r, sig);
  if (!m) {
    res.status(422).json({ ok: false, error: 'Degenerate inputs: S0, K, T and sig must be positive' });
    return;
  }
  let gridData = null;
  if (grid) {
    const n = grid.n;
    const x0 = K * 0.4, x1 = K * 1.6, y0 = 0.05, y1 = 1.6;
    const points: { x: number; y: number; z: number }[] = [];
    for (let i = 0; i < n; i++) {
      const x = x0 + (x1 - x0) * i / (n - 1);
      for (let j = 0; j < n; j++) {
        const t = y0 + (y1 - y0) * j / (n - 1);
        const mm = bs(x, K, t, r, sig);
        if (!mm) continue;
        points.push({ x, y: t, z: grid.metric === 'price' ? (opt === 'call' ? mm.C : mm.P) : (opt === 'call' ? mm.delta : mm.deltaP) });
      }
    }
    gridData = { n, metric: grid.metric, points, shape: [n, n], x_range: [x0, x1], y_range: [y0, y1] };
  }
  res.json({ opt, params: { S0, K, T, r, sig }, C: m.C, P: m.P, delta: m.delta, deltaP: m.deltaP, gamma: m.gamma, vega: m.vega, theta: m.theta, rho: m.rho, d1: m.d1, d2: m.d2, grid: gridData });
});

export default router;
